### PR TITLE
BI management cleaning

### PR DESCRIPTION
**Pull request by mohierf**
As discussed, BI (business impact) management:
- the UI has its own minimum BI (`problems_business_impact`) and the logged-in user has its own minimum BI (`min_business_impact`)
- the UI views are filtered per default with the UI minimum BI
- the search engine allows to specify BI criteria (bi>, bi<=, ...)
-----

Some cleaning in the BI management:
-when the UI and user BI are not consistent, an alert log is emitted and an information is displayed in the user preferences view
- the BI filters in the search menu are disabled if not allowed because of the configuration
- if no BI filter is specified in the search string, the default UI BI is prepended to the search query
- if several BI exist in the search query, only the last one is kept and prepended to the query

**Note** that prepending the BI in the search query will improve the search and make the BI always visible to the end user

### CODE REVIEW

File: module/module.py [hunk 1]
```diff
@@ -956,12 +957,20 @@ def check_authentication(self, username, password):
...
             self.user_info = self.auth_module.get_user_info()
             logger.info("[WebUI] User information: %s", self.user_info)
 
+            # Raise an lert if the user has a BI defined and this BI is lower than the UI's one
+            if user.min_business_impact and user.min_business_impact < self.problems_business_impact:
+                logger.warning("[WebUI] the user minimum business impact (%d) is lower than "
+                               "the UI configured business impact (%d). You should update the UI "
+                               "configuration to suit your users configuration. Some items will be "
+                               "filtered wheres the user may expect to view them!",
+                               user.min_business_impact, self.problems_business_impact)
```
Comment (maethor) [inline module/module.py:973]: I don't think we need this. The alert should be displayed to the user, but the webui administrator should not car about it.

File: module/module.py [hunk 2]
```diff
@@ -956,12 +957,20 @@ def check_authentication(self, username, password):
                 return False
 
             user = User.from_contact(c)
+            logger.info("[WebUI] User minimum business impact: %s", user.min_business_impact)
```
Comment (maethor) [inline module/module.py:960]: Should be debug.

File: module/module.py [hunk 3]
```diff
@@ -1044,7 +1053,56 @@ def get_ui_external_links(self):
...
+                    bi = int(s[2:])
+                elif s.startswith('>') or s.startswith('<') or s.startswith('='):
+                    bi = int(s[1:]) + 1
+            except ValueError:
+                pass
+            logger.debug("- search BI is >=%d", bi)
+        new_ss = ("bi:>=%d " % bi) + new_ss
+        logger.debug("- updated search string: %s", new_ss)
+
+        return new_ss
```
Comment (maethor) [inline module/module.py:1106]: This is not the right place to do this. « get_search_string » should only return the search string. You need to add all this where the search string is generated. It is too late to compute anything here, the "search_string" variable is just use as a placeholder.

File: module/plugins/problems/problems.py [hunk 4]
```diff
@@ -35,7 +35,7 @@
 
 
 def get_page():
-    app.bottle.redirect("/all?search=%s" % DEFAULT_FILTER)
+    app.bottle.redirect("/all?search=bi:>=%d %s" % (app.problems_business_impact, DEFAULT_FILTER))
```
Comment (maethor) [inline module/plugins/problems/problems.py]: Should be something like `max(app.problems_ business_impact, user.min_business_impact)` or maybe `user.min_business_impact or app.problems_business_impact` (which allows user to set less than app.problems_business_impact).

And the bi filter should be added at the end of the search, not at begining.

File: module/views/_filters.tpl [hunk 5]
```diff
@@ -26,12 +26,13 @@
...
-               <li role="presentation"><a role="menuitem" href="?search=bp:>=2">Impact : {{!helper.get_business_impact_text(2, text=True)}}</a></li>
-               <li role="presentation"><a role="menuitem" href="?search=bp:>=1">Impact : {{!helper.get_business_impact_text(1, text=True)}}</a></li>
-               <li role="presentation"><a role="menuitem" href="?search=bp:>=0">Impact : {{!helper.get_business_impact_text(0, text=True)}}</a></li>
+               %business_impact = max(user.min_business_impact, app.problems_business_impact)
+               %for idx in range(5, business_impact - 1, -1):
+               <li role="presentation"><a role="menuitem" href="?search=bi:>={{ idx }}">Impact : {{! helper.get_business_impact_text(idx, text=True)}}</a></li>
+               %end
+               %for idx in range(business_impact - 1, -1, -1):
+               <li role="presentation" class="disabled"><a role="menuitem" class="disabled" href="?search=bi:>={{ idx }}">Impact : {{! helper.get_business_impact_text(idx, text=True)}}</a></li>
+               %end
```
Comment (maethor) [inline module/views/_filters.tpl:42]: Please revert this change. I want to be able to filter on `bp:>=0` even if my `min_business_impact` is 2.

File: module/htdocs/js/shinken-actions.js [hunk 6]
```diff
@@ -599,9 +599,11 @@ $("body").on("click", ".js-recheck", function () {
     var elt = get_action_element($(this));
 
     if (elt) {
+        console.log(elt);
         recheck_now(elt);
     } else {
         $.each(selected_elements, function(idx, name){
+            console.log("Recheck selected", elt);
```
Comment (maethor) [inline module/htdocs/js/shinken-actions.js]: I don't see how this is relevant with the rest of the PR.

**Review by maethor — CHANGES_REQUESTED** [review]

File: module/module.py (hunk 1 above)
Comment (mohierf) [reply module/module.py:973]: Nope. As an admin I will be interested for such a case. Not sure I would remember the settings and this information may be interesting if a user complains.

File: module/module.py (hunk 3 above)
Comment (mohierf) [reply module/module.py:1106]: I do not agree with this. There is not only one place where the search string is built ... it may come from the form, but it can also be a direct query parameter. This functino is called when and where the search string query parameter is interesting or the applicatino. So it is, IMO, the right place to do this stuff.

File: module/views/_filters.tpl (hunk 5 above)
Comment (mohierf) [reply module/views/_filters.tpl:42]: This because you are an admin ... but a simple user sould not override what the admin configured for him/her ! So I agree to allow this menu item but if the user is an admin only. Ok with this?

File: module/htdocs/js/shinken-actions.js (hunk 6 above)
Comment (mohierf) [reply module/htdocs/js/shinken-actions.js]: Oops :wink: I will revert this

File: module/module.py (hunk 2 above)
Comment (mohierf) [reply module/module.py:960]: This log only happens on user login. I think it is better to remain at INFO level to have it in the log file.

File: module/plugins/problems/problems.py (hunk 4 above)
Comment (mohierf) [reply module/plugins/problems/problems.py]: Ok with this using the maximum of both BI and setting the filter on search end. I give a try with this to confirm it is consistent

Comment (mohierf) [reply module/plugins/problems/problems.py]: 1. Setting the BI filter in the last position is ok to me. Only two points: it is not the most effective because the BI filtering is the last filter to get applied ... and on last position the `bi>=`filter is not always visible on small screens :/

1. Using `user.min_business_impact or app.problems_business_impact` is a very good and pythonic solution :wink:

File: module/module.py (hunk 1 above)
Comment (maethor) [reply module/module.py:973]: OK

File: module/module.py (hunk 2 above)
Comment (maethor) [reply module/module.py:960]: OK

File: module/module.py (hunk 3 above)
Comment (maethor) [reply module/module.py:1106]: I keep my position, for the same reasons stated above. I will not accept this patch.

File: module/plugins/problems/problems.py (hunk 4 above)
Comment (maethor) [reply module/plugins/problems/problems.py]: Effectiveness depends on the search. On default problems search, isnot:OK will be way more discriment (then effective) than bi:>1.

And I'm fine with bi filter not always visible on small screens because it is the least important filter.

File: module/views/_filters.tpl (hunk 5 above)
Comment (maethor) [reply module/views/_filters.tpl:42]: I understand your view and I think I see what you want, but I don't agree. `min_business_impact` is just a default setting for display purpose, I don't want it to be some sort of allmighty security rule.

If you do not want users to see some services you should manage this with shinken/alignak configuration, not webui filtering rules.

**Comment by maethor** [comment]
@mohierf n'y passe pas plus de temps, je vais implémenter ce que j'ai en tête.

File: module/module.py (hunk 3 above)
Comment (mohierf) [reply module/module.py:1106]: Ok, Let's resume. The search string is bult in an HTML input field ... except when it is provided in a query parameter. So, in your opinion, all this stuff must be done in the javascript ? Else, please explain ...

File: module/plugins/problems/problems.py (hunk 4 above)
Comment (mohierf) [reply module/plugins/problems/problems.py]: Last modification is ok with this comment

File: module/views/_filters.tpl (hunk 5 above)
Comment (mohierf) [reply module/views/_filters.tpl:42]: In shinken `min_business_impact` is not a display purpose ... it is also used for the notification ways !

**Comment by maethor** [comment]
Now that I did implement what I had in mind, I propose to close this PR.
